src/recommendation/src/main/scala/SAR.scala

```diff
@@ -0,0 +1,299 @@
+      .withColumn("flatList", seqToArray(col("collect_list(itemUserAffinityPair)")))
+      .select(col(getUserCol).cast(IntegerType), col("flatList"))
+  }
+
+  /**
+    * User-to-Item affinity matrix contains for each user-item pair an affinity score of the user towards the item.
+    * Affinity score is computed as a weighted number of transactions in which the user and the item appear together,
+    * where newer transactions are weighted more than the older transactions.
+    *
+    * Diagonal elements, occ(Item i), simply represent the number of occurrences of each item. The advantage of
+    * co-occurrence is that it is very easy to update. However, it favors predictability, and the most popular items
+    * will be recommended most of the time. To alleviate that, two additional similarity measures are used: lift and
+    * Jaccard. They can be thought of as normalized co-occurrences.
+    *
+    * Lift measures how much the co-occurrence of two items is higher than it would be by chance, i.e., what is the
+    * contribution of interaction of the two items. It is obtained as
+    *
+    * lift(Item i, Item j) = cooccur(Item i, Item j) / (occ(Item i) * occ(Item j)) .
+    *
+    * Lift favors serendipity / discoverability. For example, items 2 and 5 have the same co-occurrence with item 4,
+    * but item 5 in general occurs less frequently than item 2 and will be favored by lift.
+    *
+    *
+    * Jaccard measure is defined as the number of transaction in which two items appear together divided by the
+    * number of transactions in which either of them appears:
+    *
+    * Jaccard(Item 1, Item 2) = cooccur(Item1, Item 2) / (occ(Item 1) + occ(Item 2) - cooccur(Item 1, Item 2)) .
+    *
+    * Jaccard measure is a tradeoff between co-occurrence and lift and is the default in SAR.
+    *
+    * @param dataset
+    * @return
+    */
+  private[spark] def calculateItemItemSimilarity(dataset: Dataset[_]): DataFrame = {
+
+    val spark = dataset.sparkSession
+
+    dataset.cache
+    val itemCounts = dataset
+      .groupBy(col(getItemCol))
+      .agg(countDistinct(col(getUserCol)))
+      .collect()
+      .map(r => r.get(0) -> r.getLong(1)).toMap
+
+    val broadcastItemCounts = spark.sparkContext.broadcast(itemCounts)
+
+    val calculateFeature = udf((itemID: Int, features: linalg.Vector) => {
+      val (jaccardFlag, liftFlag) =
+        if (getSimilarityFunction == "jaccard") (true, true)
+        else if (getSimilarityFunction == "lift") (false, true)
+        else (false, false)
+
+      def lift(countI: Double, countJ: Long, cooco: Double) = (cooco / (countI * countJ)).toFloat
+
+      def jaccard(countI: Double, countJ: Long, cooco: Double) = (cooco / (countI + countJ - cooco)).toFloat
+
+      val countI = features.apply(itemID)
+      features.toArray.indices.map(i => {
+        val countJ: Long = broadcastItemCounts.value.getOrElse(i, 0)
+        val cooco = features.apply(i)
+        if (!(cooco < getSupportThreshold)) {
```

Comment:
Nit for later if neccessary, you might be able to tighten this up with  higher order function. namely something like
```
val similarityLambda = getSimilarityFunction match {
  case "jaccard" => jaccard _
  case "lift" => jaccard _
  case _ => {case (_,_, cooco) => coocoo.toFloat}
}
```